### PR TITLE
877-one-to-many_swab_matching-process

### PR DESCRIPTION
**Pull request by Ohi89**
## Pre-review checklist

I have ensured that:

- [x] Code runs successfully in the development environment.
- [x] New code is tested to prove that it works. If not, justify reasoning.
- [x] New code is documented using [type hinting](https://realpython.com/lessons/type-hinting/) and following the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html) docstring format.
- [x] New code follows the project naming standards.

## Post-review checklist

As I merge these changes, I have:

- [ ] Updated the related Jira ticket, including approximate time spent (including review).
- [ ] Moved the related Jira ticket to Done.

### CODE REVIEW

**Comment by Ohi89** [comment]
- Delete internal flag column names
- replace UDF for one of the steps
- replace hard-coded column name
- Tidy up the docstring column
- Change count_blood to count_iqvia

File: cishouseholds/merge.py [hunk 1]
```diff
@@ -438,3 +427,187 @@ def many_to_many_flag(
...
+    ----------
+    df
+    out_of_date_range_flag
+        Column to work out whether the date difference is within an upper/lower range. This column can be
+        created by assign_time_difference_and_flag_if_outside_interval() by inputing two date columns.
+    count_barcode_labs_column_name
+        the Many column from One-to-Many
+    count_barcode_voyager_column_name
+        the One column from One-to-Many
+    window_column
```
Comment (osborj) [inline cishouseholds/merge.py:470]: group_by_column?

File: cishouseholds/merge.py [hunk 2]
```diff
@@ -438,3 +427,187 @@ def many_to_many_flag(
...
+        the barcode, user_id used for grouping for the steps 2, 3, 4 where window function is used
+    ordering_columns
+        a list of strings with the column name for ordering used in steps 2 and 4.
+    mk_result_column_name
```
Comment (osborj) [inline cishouseholds/merge.py:474]: Change reference to 'mk_result'

Comment (osborj) [reply cishouseholds/merge.py:474]: pcr_result_column_name?

File: cishouseholds/merge.py [hunk 3]
```diff
@@ -438,3 +427,187 @@ def many_to_many_flag(
...
+    combination_flag_column_name
```
Comment (osborj) [inline cishouseholds/merge.py:475]: 'combination' probably not needed - flag_column_name, column_name_to_assign, etc

File: cishouseholds/merge.py [hunk 4]
```diff
@@ -438,3 +427,187 @@ def many_to_many_flag(
...
+        Output column in Step 2
+    """
+    window = Window.partitionBy(window_column).orderBy(*ordering_columns)
+    return df.withColumn(time_order_logic_flag_column_name, F.rank().over(window)).withColumn(
+        time_order_logic_flag_column_name,
+        F.when(F.col(time_order_logic_flag_column_name) == 1, None).otherwise(1),
+    )
+
+
+def merge_one_to_many_swab_result_mk_logic(
```
Comment (osborj) [inline cishouseholds/merge.py:545]: 'mk'

File: cishouseholds/merge.py [hunk 5]
```diff
@@ -438,3 +427,187 @@ def many_to_many_flag(
...
+    df: DataFrame,
+    void_value: Union[str, int],
+    window_column: str,
+    mk_result_column_name: str,
+    result_mk_logic_flag_column_name: str,
+) -> DataFrame:
+    """
+    Step 3: drops result mk if there are void values having at least a positive/negative within the
```
Comment (osborj) [inline cishouseholds/merge.py:553]: here

File: cishouseholds/merge.py [hunk 6]
```diff
@@ -438,3 +427,187 @@ def many_to_many_flag(
...
+        "merge_flag",
+        F.when(F.col("merge_flag") == 1, None).otherwise(1),
+    )
+    df = merge_one_to_many_swab_time_date_logic(df, window_column, ordering_columns, "time_order_flag")
+    df = merge_one_to_many_swab_result_mk_logic(
+        df=df,
+        void_value=void_value,
+        window_column=window_column,
+        mk_result_column_name=mk_result_column_name,
+        result_mk_logic_flag_column_name="mk_flag",
```
Comment (osborj) [inline cishouseholds/merge.py:501]: pcr_result_classification

**Review by foster999 — APPROVED** [review]

**Pull request merged**